crates/services/p2p/src/codecs/postcard.rs

```diff
@@ -160,30 +176,200 @@ impl GossipsubCodec for PostcardCodec {
+            ResponseMessageErrorCode::ProtocolV1EmptyResponse,
+        ));
+        let mut codec = PostcardCodec::new(1024);
+        let mut buf = Vec::with_capacity(1024);
+        codec
+            .write_response(&PostcardProtocol::V2, &mut buf, response.clone())
+            .await
+            .expect("Valid Vec<SealedBlockHeader> is serialized using v1");
+
+        let deserialized = codec
+            .read_response(&PostcardProtocol::V2, &mut buf.as_slice())
+            .await
+            .expect("Valid Vec<SealedBlockHeader> is deserialized using v1");
+
+        match deserialized {
+            ResponseMessage::SealedHeaders(Err(
+                ResponseMessageErrorCode::ProtocolV1EmptyResponse,
+            )) => {}
+            other => {
+                panic!("Deserialized to {other:?}, expected {response:?}")
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn serialzation_roundtrip_using_v1_error_response() {
+        let response = ResponseMessage::SealedHeaders(Err(
+            ResponseMessageErrorCode::ProtocolV1EmptyResponse,
+        ));
+        let mut codec = PostcardCodec::new(1024);
+        let mut buf = Vec::with_capacity(1024);
+        codec
+            .write_response(&PostcardProtocol::V1, &mut buf, response.clone())
+            .await
+            .expect("Valid Vec<SealedBlockHeader> is serialized using v1");
+
+        let deserialized = codec
+            .read_response(&PostcardProtocol::V1, &mut buf.as_slice())
+            .await
+            .expect("Valid Vec<SealedBlockHeader> is deserialized using v1");
+
+        match deserialized {
+            ResponseMessage::SealedHeaders(Err(
+                ResponseMessageErrorCode::ProtocolV1EmptyResponse,
+            )) => {}
+            other => {
+                panic!("Deserialized to {other:?}, expected {response:?}")
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn backward_compatibility_v1_read_error_response() {
+        let response = ResponseMessage::SealedHeaders(Err(
+            ResponseMessageErrorCode::ProtocolV1EmptyResponse,
+        ));
+        let mut codec = PostcardCodec::new(1024);
+        let mut buf = Vec::with_capacity(1024);
+        codec
+            .write_response(&PostcardProtocol::V1, &mut buf, response.clone())
+            .await
+            .expect("Valid Vec<SealedBlockHeader> is serialized using v1");
+
+        let deserialized_as_legacy =
+            // We cannot access the codec trait from an old node here, 
+            // so we deserialize directly using the `LegacyResponseMessage` type.
+            deserialize::<LegacyResponseMessage>(&buf).expect("Deserialization as LegacyResponseMessage should succeed");
+        match deserialized_as_legacy {
+            LegacyResponseMessage::SealedHeaders(None) => {}
+            other => {
+                panic!("Deserialized to {other:?}, expected {response:?}")
+            }
+        }
+    }
```

Comment:
Another example of how to structure the tests:
```suggestion
    #[tokio::test]
    async fn codec__write_response_is_backwards_compatible_with_v1() {
        // given
        let response = ResponseMessage::SealedHeaders(Err(
            ResponseMessageErrorCode::ProtocolV1EmptyResponse,
        ));
        let mut codec = PostcardCodec::new(1024);
        let mut buf = Vec::with_capacity(1024);
        
        // when
        codec
            .write_response(&PostcardProtocol::V1, &mut buf, response.clone())
            .await
            .expect("Valid Vec<SealedBlockHeader> is serialized using v1");
            
        let deserialized_as_legacy =
            // We cannot access the codec trait from an old node here, 
            // so we deserialize directly using the `LegacyResponseMessage` type.
            deserialize::<LegacyResponseMessage>(&buf).expect("Deserialization as LegacyResponseMessage should succeed");
            
        // then
        match deserialized_as_legacy {
            LegacyResponseMessage::SealedHeaders(None) => {}
            other => {
                panic!("Deserialized to {other:?}, expected {response:?}")
            }
        }
    }
```

Reply:
I have changed the names of the tests to follow your suggestions in 6276af58621d120f5b4e4526b5c0d2fa19f93f58